File: solutions/trees/105-construct-binary-tree-from-preorder-and-inorder-traversal.py
"""
# Difficulty: Medium

# 105. Construct Binary Tree From Preorder And Inorder Traversal

Given two integer arrays preorder and inorder where preorder is the preorder traversal of a binary tree and inorder is the inorder traversal of the same tree, construct and return the binary tree.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>preorder = [3,9,20,15,7], inorder = [9,3,15,20,7]</dd>
<dt>Output:</dt>
<dd>Tree with root 3</dd>
<dt>Explanation:</dt>
<dd>The tree is uniquely constructed from preorder [3,9,20,15,7] and inorder [9,3,15,20,7] traversals</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
Recursively build tree. Find root in preorder (first element). Find root in inorder (splits left/right). Recursively build left subtree with elements before root, right subtree with elements after root in inorder.

### APPROACH:
1. **Build index map**: Create dict mapping inorder values to indices
2. **Define helper**: Implement build(pre_start, pre_end, in_start, in_end)
3. **Base case**: If pre_start > pre_end, return None
4. **Create root**: root = TreeNode(preorder[pre_start])
5. **Find root in inorder**: root_idx = inorder_map[root.val]
6. **Calculate left size**: left_size = root_idx - in_start
7. **Build left subtree**: root.left = build(pre_start+1, pre_start+left_size, in_start, root_idx-1)
8. **Build right subtree**: root.right = build(pre_start+left_size+1, pre_end, root_idx+1, in_end), return root

### WHY THIS WORKS:
- Preorder gives root (first element), inorder splits left/right subtrees
- Find root in inorder: elements left of root are left subtree, right are right subtree
- Recursively build left and right subtrees with their preorder/inorder slices
- Hash map stores inorder indices for O(1) root lookup instead of O(n) search
- O(n) time: each node processed once, O(n) space for map and recursion

### EXAMPLE WALKTHROUGH:
```
Input: preorder = [3,9,20,15,7], inorder = [9,3,15,20,7]
Step 1: Root is first in preorder
  root = 3

Step 2: Find root in inorder
  left subtree: [9]
  right subtree: [15,20,7]

Step 3: Recursively build
  left: preorder=[9], inorder=[9]
  right: preorder=[20,15,7], inorder=[15,20,7]

Output: Tree with root 3
```

### TIME COMPLEXITY:
O(n)

### SPACE COMPLEXITY:
O(1)

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>
"""

from typing import List, Optional, Dict, Tuple

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def constructFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        """
        Constructs a binary tree from its preorder traversal.
        
        Args:
            preorder: List of integers representing preorder traversal of the tree
            
        Returns:
            Root node of the constructed binary tree
        """
        if not preorder:
            return None
        
        self.preorder_index = 0
        
        def construct(bound: int) -> Optional[TreeNode]:
            """
            Helper function to construct the tree recursively.
            
            Args:
                bound: Upper bound value for the current subtree
                
            Returns:
                Root node of the current subtree
            """
            if (self.preorder_index >= len(preorder) or 
                preorder[self.preorder_index] > bound):
                return None
            
            # Create root node with current value
            root = TreeNode(preorder[self.preorder_index])
            self.preorder_index += 1
            
            # Recursively construct left and right subtrees
            # Left subtree values must be less than root's value
            root.left = construct(root.val)
            # Right subtree values must be less than bound
            root.right = construct(bound)
            
            return root
        
        return construct(float('inf'))

    def printTree(self, root: Optional[TreeNode]) -> None:
        """
        Prints the tree in a level-order traversal.
        
        Args:
            root: Root node of the binary tree
        """
        if not root:
            print("Empty tree")
            return
        
        queue = [root]
        while queue:
            level = []
            level_size = len(queue)
            
            for _ in range(level_size):
                node = queue.pop(0)
                if node:
                    level.append(str(node.val))
                    queue.append(node.left)
                    queue.append(node.right)
                else:
                    level.append("null")
                    
            # Remove trailing nulls
            while level and level[-1] == "null":
                level.pop()
                
            if level:
                print(" ".join(level))

def test_solution():
    """
    Test cases for the solution.
    """
    solution = Solution()

    # Test case 1: Basic case
    result = solution.constructFromPreorder([1, 2, 3])
    expected = <solution_module.TreeNode object at 0x101214ec0>
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Empty input
    result = solution.constructFromPreorder([])
    expected = None
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 3: Single element
    result = solution.constructFromPreorder([1])
    expected = <solution_module.TreeNode object at 0x101225480>
    assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 105. Construct Binary Tree From Preorder And Inorder Traversal")

File: solutions/trees/alternatives/105-construct-binary-tree-from-preorder-and-inorder-traversal.js
/**

 *
 * This problem demonstrates key concepts in tree construction from traversals.
 *
 * SOLUTION EXPLANATION:
 *
 * INTUITION:
 * Preorder gives us the root first, then left subtree, then right subtree.
 * Inorder gives us left subtree, then root, then right subtree.
 * By combining these patterns, we can uniquely reconstruct the binary tree.
 * The first element in preorder is always the root. Find this root in inorder
 * to determine left and right subtrees, then recursively build each subtree.
 *
 * APPROACH:




 *
 * WHY THIS WORKS:
 * - Preorder traversal visits root before subtrees (Root-Left-Right)
 * - Inorder traversal visits left subtree, then root, then right (Left-Root-Right)
 * - Root position in inorder divides tree into left and right parts
 * - Recursive structure naturally builds the tree from top to bottom
 *
 * TIME COMPLEXITY: O(n) where n is number of nodes
 * Each node is processed once, hashmap lookups are O(1)
 *
 * SPACE COMPLEXITY: O(n) for hashmap + O(h) for recursion stack
 * Where h is height of tree (worst case O(n) for skewed tree)
 *
 * EXAMPLE WALKTHROUGH:
 * ```
 * Input: preorder = [3,9,20,15,7], inorder = [9,3,15,20,7]
 *
 * Step 1: Root = 3 (first in preorder)
 * Step 2: Find 3 in inorder at index 1
 *         Left subtree: [9], Right subtree: [15,20,7]
 * Step 3: Recursively build:
 *         - Left: preorder=[9], inorder=[9] → node 9
 *         - Right: preorder=[20,15,7], inorder=[15,20,7]
 *           - Root = 20, Left=[15], Right=[7]
 *
 * Output:     3
 *            / \
 *           9  20
 *             /  \
 *            15   7
 * ```
 *
 * EDGE CASES:
 * - Empty arrays (return null)
 * - Single element (return single node)
 * - Left-skewed or right-skewed trees
 * - All elements unique (guaranteed by problem)
 */

/**
 * Definition for a binary tree node.
 */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}

/**
 * Main solution for Problem 105: Construct Binary Tree From Preorder And Inorder Traversal
 *
 * @param {number[]} preorder - Preorder traversal array
 * @param {number[]} inorder - Inorder traversal array
 * @return {TreeNode} - Root of constructed binary tree
 *
 * Time Complexity: O(n) where n is number of nodes
 * Space Complexity: O(n) for hashmap + O(h) for recursion
 */
function solve(preorder, inorder) {
    return buildTree(preorder, inorder);
}

/**
 * Approach 1: Optimized with HashMap for O(1) inorder lookups
 *
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
function buildTree(preorder, inorder) {
    if (!preorder || !inorder || preorder.length === 0 || inorder.length === 0) {
        return null;
    }

    // Build hashmap for O(1) inorder index lookups
    const inorderMap = new Map();
    for (let i = 0; i < inorder.length; i++) {
        inorderMap.set(inorder[i], i);
    }

    let preorderIndex = 0;

    function buildTreeRecursive(inorderStart, inorderEnd) {
        if (inorderStart > inorderEnd) {
            return null;
        }

        // Root is the current element in preorder
        const rootVal = preorder[preorderIndex++];
        const root = new TreeNode(rootVal);

        // Find root position in inorder
        const rootIndex = inorderMap.get(rootVal);

        // Build left subtree first (preorder processes left before right)
        root.left = buildTreeRecursive(inorderStart, rootIndex - 1);
        // Build right subtree
        root.right = buildTreeRecursive(rootIndex + 1, inorderEnd);

        return root;
    }

    return buildTreeRecursive(0, inorder.length - 1);
}

/**
 * Approach 2: Simple recursive without hashmap optimization
 *
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
function buildTreeSimple(preorder, inorder) {
    if (!preorder || !inorder || preorder.length === 0 || inorder.length === 0) {
        return null;
    }

    // Root is first element in preorder
    const rootVal = preorder[0];
    const root = new TreeNode(rootVal);

    // Find root in inorder
    const rootIndex = inorder.indexOf(rootVal);

    // Split arrays for left and right subtrees
    const leftPreorder = preorder.slice(1, rootIndex + 1);
    const leftInorder = inorder.slice(0, rootIndex);
    const rightPreorder = preorder.slice(rootIndex + 1);
    const rightInorder = inorder.slice(rootIndex + 1);

    // Recursively build subtrees
    root.left = buildTreeSimple(leftPreorder, leftInorder);
    root.right = buildTreeSimple(rightPreorder, rightInorder);

    return root;
}

/**
 * Approach 3: Iterative solution using stack
 *
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
function buildTreeIterative(preorder, inorder) {
    if (!preorder || preorder.length === 0) {
        return null;
    }

    const root = new TreeNode(preorder[0]);
    const stack = [root];
    let preIndex = 1;
    let inIndex = 0;

    while (preIndex < preorder.length) {
        let node = stack[stack.length - 1];

        if (node.val !== inorder[inIndex]) {
            // Create left child
            node.left = new TreeNode(preorder[preIndex++]);
            stack.push(node.left);
        } else {
            // Pop nodes that match inorder sequence
            while (stack.length > 0 && stack[stack.length - 1].val === inorder[inIndex]) {
                node = stack.pop();
                inIndex++;
            }
            // Create right child
            node.right = new TreeNode(preorder[preIndex++]);
            stack.push(node.right);
        }
    }

    return root;
}

/**
 * Helper function to convert tree to array for testing
 * @param {TreeNode} root
 * @return {number[]}
 */
function treeToArray(root) {
    if (!root) return [];

    const result = [];
    const queue = [root];

    while (queue.length > 0) {
        const node = queue.shift();
        if (node) {
            result.push(node.val);
            queue.push(node.left);
            queue.push(node.right);
        } else {
            result.push(null);
        }
    }

    // Remove trailing nulls
    while (result.length > 0 && result[result.length - 1] === null) {
        result.pop();
    }

    return result;
}

/**
 * Test cases for Problem 105: Construct Binary Tree From Preorder And Inorder Traversal
 */
function testSolution() {
    console.log('Testing 105. Construct Binary Tree From Preorder And Inorder Traversal');

    // Test case 1: Standard tree
    const preorder1 = [3, 9, 20, 15, 7];
    const inorder1 = [9, 3, 15, 20, 7];
    const result1 = solve(preorder1, inorder1);
    const expected1 = [3, 9, 20, null, null, 15, 7];
    const actual1 = treeToArray(result1);
    console.assert(JSON.stringify(actual1) === JSON.stringify(expected1),
        `Test 1 failed: expected ${JSON.stringify(expected1)}, got ${JSON.stringify(actual1)}`);

    // Test case 2: Single node
    const preorder2 = [1];
    const inorder2 = [1];
    const result2 = solve(preorder2, inorder2);
    const expected2 = [1];
    const actual2 = treeToArray(result2);
    console.assert(JSON.stringify(actual2) === JSON.stringify(expected2),
        `Test 2 failed: expected ${JSON.stringify(expected2)}, got ${JSON.stringify(actual2)}`);

    // Test case 3: Empty arrays
    const preorder3 = [];
    const inorder3 = [];
    const result3 = solve(preorder3, inorder3);
    console.assert(result3 === null,
        `Test 3 failed: expected null, got ${result3}`);

    // Test case 4: Left-skewed tree
    const preorder4 = [1, 2, 3];
    const inorder4 = [3, 2, 1];
    const result4 = solve(preorder4, inorder4);
    const expected4 = [1, 2, null, 3];
    const actual4 = treeToArray(result4);
    console.assert(JSON.stringify(actual4) === JSON.stringify(expected4),
        `Test 4 failed: expected ${JSON.stringify(expected4)}, got ${JSON.stringify(actual4)}`);

    // Test case 5: Right-skewed tree
    const preorder5 = [1, 2, 3];
    const inorder5 = [1, 2, 3];
    const result5 = solve(preorder5, inorder5);
    const expected5 = [1, null, 2, null, 3];
    const actual5 = treeToArray(result5);
    console.assert(JSON.stringify(actual5) === JSON.stringify(expected5),
        `Test 5 failed: expected ${JSON.stringify(expected5)}, got ${JSON.stringify(actual5)}`);

    // Test case 6: Perfect binary tree
    const preorder6 = [1, 2, 4, 5, 3, 6, 7];
    const inorder6 = [4, 2, 5, 1, 6, 3, 7];
    const result6 = solve(preorder6, inorder6);
    const expected6 = [1, 2, 3, 4, 5, 6, 7];
    const actual6 = treeToArray(result6);
    console.assert(JSON.stringify(actual6) === JSON.stringify(expected6),
        `Test 6 failed: expected ${JSON.stringify(expected6)}, got ${JSON.stringify(actual6)}`);

    // Test alternative approaches give same results
    const resultSimple = buildTreeSimple(preorder1, inorder1);
    const actualSimple = treeToArray(resultSimple);
    console.assert(JSON.stringify(actualSimple) === JSON.stringify(expected1),
        'Simple approach should give same result');

    const resultIterative = buildTreeIterative(preorder1, inorder1);
    const actualIterative = treeToArray(resultIterative);
    console.assert(JSON.stringify(actualIterative) === JSON.stringify(expected1),
        'Iterative approach should give same result');

    console.log('All test cases passed for 105. Construct Binary Tree From Preorder And Inorder Traversal!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 105. Construct Binary Tree From Preorder And Inorder Traversal ===');
    console.log('Category: Trees');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    buildTree,
    buildTreeSimple,
    buildTreeIterative,
    TreeNode,
    treeToArray,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - Three different approaches with different trade-offs:



 * - Key insight: preorder gives root, inorder splits left/right subtrees
 * - HashMap optimization crucial for large inputs to avoid O(n²) behavior
 * - Recursive approach naturally follows the tree construction pattern
 * - All approaches handle edge cases consistently
 * - Problem guarantees unique values, making solution deterministic
 */
